File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.js
function solution(picks, minerals) {
    let answer = 0;
    let sum = 0;
    for (let i = 0; i < picks.length; i++) sum += picks[i];
    
    let m = [];
    for (let i = 0; i < sum; i++) m.push(minerals.splice(0, 5));
    
    m.sort((a, b) => {
        let diamondCountA = a.filter(mineral => mineral === 'diamond').length;
        let diamondCountB = b.filter(mineral => mineral === 'diamond').length;
        let ironCountA = a.filter(mineral => mineral === 'iron').length;
        let ironCountB = b.filter(mineral => mineral === 'iron').length;
        let stoneCountA = a.filter(mineral => mineral === 'stone').length;
        let stoneCountB = b.filter(mineral => mineral === 'stone').length;

        if (diamondCountA === diamondCountB) {
            if (ironCountA === ironCountB) {
                return stoneCountB - stoneCountA;
            }
            return ironCountB - ironCountA;
        }
        return diamondCountB - diamondCountA;
    });
    
    for (let i = 0; i < picks.length; i++) {
        let count = picks[i];
        let n1, n2, n3;
        
        if (i === 0) {
            n1 = 1;
            n2 = 1;
            n3 = 1;
        } else if (i === 1) {
            n1 = 5;
            n2 = 1;
            n3 = 1;
        } else {
            n1 = 25;
            n2 = 5;
            n3 = 1;
        }
        
        while (count !== 0) {
            if (m.length === 0) return answer;
            
            let diamondCount = 0,
                ironCount = 0,
                stoneCount = 0;
            
            for (let j = 0; j < m[0].length; j++) {
                if (m[0][j] === 'diamond') diamondCount++;
                else if (m[0][j] === 'iron') ironCount++;
                else if (m[0][j] === 'stone') stoneCount++;
            }
            
            answer += diamondCount * n1 + ironCount * n2 + stoneCount * n3;
            count--;
            
            m.shift();
        }
    }
    return answer;
}

File: Programmers/2/172927. 광물 캐기/광물 캐기.java
import java.util.*;

class Solution {
    public int solution(int[] picks, String[] minerals) {
        int answer = 0;
        
        int index = Math.min(minerals.length / 5 + 1, picks[0] + picks[1] + picks[2]);
        int[][] sum = new int[index][3];
        int dPoint = 0, iPoint = 0, sPoint = 0;
        
        for(int i = 0; i < minerals.length; i += 5) {
            if(i / 5 == index) break;
            
            for(int j = i; j < i+5; j++) {
                String m = minerals[j];
                
                if(m.equals("diamond")) {
                    dPoint += 1;
                    iPoint += 5;
                    sPoint += 25;
                } else if(m.equals("iron")) {
                    dPoint += 1;
                    iPoint += 1;
                    sPoint += 5;
                } else {
                    dPoint += 1;
                    iPoint += 1;
                    sPoint += 1;
                }
                
                if(j == minerals.length - 1) break;
            }
            
            sum[i/5][0] += dPoint;
            sum[i/5][1] += iPoint;
            sum[i/5][2] += sPoint;
            
            dPoint = iPoint = sPoint = 0;
        }
        
        Arrays.sort(sum, (o1, o2) -> o2[2] - o1[2]);
        
        for(int i = 0; i < index; i++) {
            if(picks[0] != 0) {
                answer += sum[i][0];
                picks[0]--;
            } else if(picks[1] != 0) {
                answer += sum[i][1];
                picks[1]--;
            } else {
                answer += sum[i][2];
                picks[2]--;
            }
        }
        
        return answer;
    }
}